### PR TITLE
Slurm image checks and Documentation Updates

### PR DESCRIPTION
**Pull request by cdunbar13**
Updates to check the instance_image family and project settings for slurm deployments.  Tied to v5.7 for compatibility, but can be expanded if needed.

Updates to documentation as well.

### CODE REVIEW

File: community/modules/compute/schedmd-slurm-gcp-v5-node-group/source_image_logic.tf
Comment (tpdownes) [inline community/modules/compute/schedmd-slurm-gcp-v5-node-group/source_image_logic.tf:39]: I believe that we want to distinguish between 2 situations:

1. The blueprint is using the schedmd-slurm-public project but an incompatible image.
2. The blueprint is using another project entirely

I'm also concerned about throwing the words "not supported" around when we really mean "incompatible." Let's discuss specific wording offline.

I think you'll want two postconditions for this and it will probably look something like below. The error messages are not intended to be taken verbatim, but an idea of how to distinguish between the different scenarios.

```
  condition =  var.instance_image_override || self.project == "schedmd-slurm-public"
  error_message = "If using images published outside the schedmd-slurm-public project, you must set var.instance_image_override to true and build images according to LINK-TO-URL."
```

The OR condition could also be a `contains(keys...`. A second postcondition might look like

```
  condition =  self.project != "schedmd-slurm-public" || try(contains(local.supported_project_families[self.project], self.family), false)
  error_message = "If using images published in the schedmd-slurm-public project, you must use one of these compatible releases: X,Y,Z"
```

I don't think we want to get in the business of supporting alternative families within the schedmd project.

File: community/modules/scheduler/schedmd-slurm-gcp-v5-controller/variables.tf
Comment (tpdownes) [inline community/modules/scheduler/schedmd-slurm-gcp-v5-controller/variables.tf:553]: 1. Let's discuss offline "override" as the terminology to use. I like having the variable being lexically close to "instance_image."
2. I think the text should steer the user toward a statement like "the image must be built to be compatible with release according to LINK-TO-URL" rather than "custom image that is not supported".

**Review by tpdownes — COMMENTED** [review]
Initial round of feedback.

File: community/modules/scheduler/schedmd-slurm-gcp-v5-login/variables.tf
Comment (cdunbar13) [reply community/modules/scheduler/schedmd-slurm-gcp-v5-login/variables.tf]: > I think we should discuss this variable name and description a bit offline:
> 
> 1. I like that it is lexically next to instance_image so that it appears that way in documentation. I think I might choose a different word than override. Let's think a bit.

I believe I've gotten a bit closer to your intention with for this.  I changed the variable to `instance_image_compatible`, but we can discuss.

> 2. I think we should use different language than "not supported by" and more like "and must be produced with same release of Terraform module according to docs at https://X"

Understood, I believe the new language covers this in both the errors messages, variable descriptions, and documentation.

File: community/modules/compute/schedmd-slurm-gcp-v5-node-group/source_image_logic.tf
Comment (cdunbar13) [reply community/modules/compute/schedmd-slurm-gcp-v5-node-group/source_image_logic.tf:39]: I've updated the language (but we should still discuss), and took a different approach for the two potential errors.  Instead of two postconditions, which may both trigger, I've created one postcondition with a conditional for the error message.  It should give a more accurate error message depending on if the user has specified an incorrect project or image family.

File: community/modules/scheduler/schedmd-slurm-gcp-v5-controller/variables.tf
Comment (cdunbar13) [reply community/modules/scheduler/schedmd-slurm-gcp-v5-controller/variables.tf]: Same as above.

File: community/modules/compute/schedmd-slurm-gcp-v5-node-group/source_image_logic.tf
Comment (tpdownes) [reply community/modules/compute/schedmd-slurm-gcp-v5-node-group/source_image_logic.tf:39]: I do not see how my two postconditions can simultaneously trigger. Can you demonstrate a combination of settings that does?

My suggestion seems very direct to read and the error messages are a lot shorter. I want to get these error messages really down to "Here is your immediate problem. Visit URL for details." Preferably one line.

File: community/modules/compute/schedmd-slurm-gcp-v5-node-group/variables.tf
Comment (tpdownes) [inline community/modules/compute/schedmd-slurm-gcp-v5-node-group/variables.tf:115]: I suggest:

> that the user has built or copied a custom image that meets compatibility requirements defined at URL.
> This setting is ignored if using official images built by SchedMD

File: docs/vm-images.md
Comment (tpdownes) [inline docs/vm-images.md:221]: Misspelling: "Publicly"

File: docs/vm-images.md
Comment (tpdownes) [inline docs/vm-images.md:19]: I'm not convinced that this new section is helping this PR especially as we have not yet harmonized the dtype of instance_image into an object of defined fields.

File: docs/vm-images.md
Comment (tpdownes) [inline docs/vm-images.md:223]: I think you need to draft text here that is crystal clear that the Terraform module version and Packer module version must be the same. That should be the first sentence. The second (and third?) sentences should be that SchedMD publishes these images and that our modules will enforce compatibility when using images published by SchedMD.

The next paragaph should cover custom images. We can review that text in next round.

**Review by tpdownes — CHANGES_REQUESTED** [review]
Make sure that we are using the data source as inputs to module where we can. We should have discussion with teammates about the project normalized value as that is funny, but (as I recall) for a good reason.

File: community/modules/compute/schedmd-slurm-gcp-v5-node-group/source_image_logic.tf
Comment (cdunbar13) [reply community/modules/compute/schedmd-slurm-gcp-v5-node-group/source_image_logic.tf]: Done.

File: community/modules/compute/schedmd-slurm-gcp-v5-node-group/source_image_logic.tf
Comment (cdunbar13) [reply community/modules/compute/schedmd-slurm-gcp-v5-node-group/source_image_logic.tf:25]: Done.

File: community/modules/compute/schedmd-slurm-gcp-v5-node-group/main.tf
Comment (cdunbar13) [reply community/modules/compute/schedmd-slurm-gcp-v5-node-group/main.tf]: Resolved

File: community/modules/compute/schedmd-slurm-gcp-v5-node-group/source_image_logic.tf
Comment (tpdownes) [inline community/modules/compute/schedmd-slurm-gcp-v5-node-group/source_image_logic.tf:20]: Let's have the formatting align with rest of Toolkit choices. Drop the quotation marks around the key (it can only be a string in Terraform) and separate each value by line. e.g.

```hcl
known_project_families = {
  schedmd-slurm-public = [
    "slurm-gcp-5-7-debian-11",
    ...
  ]
}
```

File: community/modules/compute/schedmd-slurm-gcp-v5-node-group/source_image_logic.tf
Comment (tpdownes) [inline community/modules/compute/schedmd-slurm-gcp-v5-node-group/source_image_logic.tf:45]: you do not need to wrap this in a `try` function.

File: community/modules/compute/schedmd-slurm-gcp-v5-node-group/source_image_logic.tf
Comment (tpdownes) [inline community/modules/compute/schedmd-slurm-gcp-v5-node-group/source_image_logic.tf:51]: I don't think `var.instance_image_custom` should appear in this condition. The logic is that the first condition will fire if you aren't using the schedmd-slurm-public project (in which case you must declare your intent to use a custom image) and the 2nd condition will fire if you are using schedmd-slurm-public but are using an incompatible family. There is not a situation where we want people to be able to override that logic within schedmd-slurm-public.

Comment (tpdownes) [reply community/modules/compute/schedmd-slurm-gcp-v5-node-group/source_image_logic.tf:51]: I think the logic reduces to:

```hcl
!contains(keys(local.known_project_families), self.project) || try(contains(local.known_project_families[self.project], self.family)
```

If Terraform did short-circuiting on booleans, you wouldn't need the `try`. But it doesn't.

**Review by tpdownes — CHANGES_REQUESTED** [review]
I think we are close. If you agree with these changes, make them and then squash some of the back/forth so that the git history isn't so cluttered. After that, I'll try to break it with different combinations of project/name/family.

File: community/modules/compute/schedmd-slurm-gcp-v5-node-group/source_image_logic.tf
Comment (cdunbar13) [reply community/modules/compute/schedmd-slurm-gcp-v5-node-group/source_image_logic.tf:51]: I'll make the fix removing the first condition, but it also makes it impossible for users to utilize (at their own peril) other schedmd images with the wrapper modules for v5.  I'd argue that `instance_image_custom: true` already lets them use any image, so why not other schedmd images?

File: community/modules/compute/schedmd-slurm-gcp-v5-node-group/source_image_logic.tf
Comment (cdunbar13) [reply community/modules/compute/schedmd-slurm-gcp-v5-node-group/source_image_logic.tf]: Resolved.

File: community/modules/compute/schedmd-slurm-gcp-v5-node-group/source_image_logic.tf
Comment (cdunbar13) [reply community/modules/compute/schedmd-slurm-gcp-v5-node-group/source_image_logic.tf]: Resolved.

File: community/modules/compute/schedmd-slurm-gcp-v5-node-group/source_image_logic.tf
Comment (cdunbar13) [reply community/modules/compute/schedmd-slurm-gcp-v5-node-group/source_image_logic.tf:49]: Resolved.

File: community/modules/scheduler/schedmd-slurm-gcp-v5-login/variables.tf
Comment (cdunbar13) [reply community/modules/scheduler/schedmd-slurm-gcp-v5-login/variables.tf]: Resolved.

File: community/modules/compute/schedmd-slurm-gcp-v5-node-group/source_image_logic.tf
Comment (tpdownes) [inline community/modules/compute/schedmd-slurm-gcp-v5-node-group/source_image_logic.tf:53]: In interest of closing loop more quickly, I suggest that you modify this message to read:

```
      Images in project ${self.project} are not published by SchedMD. Images must be
      built by compatible releases of the Terraform and Packer modules following the
      guidance at https://goo.gle/hpc-slurm-images. Set var.instance_image_custom to
      true to silence this error and acknowledge that you are using a compatible image.     
```

File: community/modules/compute/schedmd-slurm-gcp-v5-node-group/variables.tf
Comment (tpdownes) [inline community/modules/compute/schedmd-slurm-gcp-v5-node-group/variables.tf:123]: Modify to https://goo.gle/hpc-slurm-images

File: docs/vm-images.md
Comment (tpdownes) [inline docs/vm-images.md:21]: I don't think we should include this change in this PR because it's really unrelated. But I will defer to @cboneti who plans a follow-up review for this PR.

File: docs/vm-images.md
Comment (tpdownes) [inline docs/vm-images.md:238]: I would reduce the rest of the text to provide 2 links to `examples/README.md` with the anchors for image-builder.yaml and ml-slurm.yaml.

File: docs/vm-images.md
Comment (tpdownes) [inline docs/vm-images.md:227]: I would rather not name a specific release in the text especially as this only matches the beginning of each families name. Suggest removing the line in favor of:

```
Slurm images are compatible by the minor version releases of the Terraform and Packer modules. For example, images built for version 5.8 are compatible with all 5.8 Terraform modules beginning with 5.8.0. The version of the Slurm modules used by your copy of the Toolkit in the local filesystem can be inspected by looking for the `source` line in community/modules/compute/schedmd-slurm-gcp-v5-partition/main.tf.
```

The latest GitHub release supports [these images](https://github.com/SchedMD/slurm-gcp/blob/5.7.6/docs/images.md#published-image-family).

@cboneti This is definitely open for some thought.

**Review by tpdownes — CHANGES_REQUESTED** [review]
This is my last review. I think we're at a place where the logic is good and the final text needs some massaging. My suggestions, frankly, reverse a bit of the "keep it super short" requests, so accept my apologies but I think they are clearer.

In addition to the changes, I've requested, please:

1. Correct a bunch of trailing newlines here and there.
1. Accept feedback from @cboneti 
1. Wait for #1728 to be merged, rebase off develop and update any explicit references from 5.7.6 to 5.8.0.
1. Update examples/image-builder.yaml and examples/ml-slurm.yaml to include the necessary extra setting.

**Review by tpdownes — COMMENTED** [review]
LGTM. @cboneti requested an overall review when I was done.

**Review by cboneti — CHANGES_REQUESTED** [review]
I left a few comments

File: docs/vm-images.md
Comment (cdunbar13) [reply docs/vm-images.md:219]: For now we could probably point to their version branch (v5) instead of the tag.  It would present the same problem as before though, where they go ahead a version and it takes us a few days/weeks/months to update to that version.  That's what I'm guessing happened with the person that was wondering why the 6.1 image didn't work (we were pointing to the Slurm-GCP master branch).  

This problem seems related but it's not immediately apparent to me how a solution would solve all of the versioning issues.  I will try and think on it.  Maybe this is another task.

Comment (cdunbar13) [reply docs/vm-images.md:292]: Switched to v5, but it will have the issues mentioned above.

File: community/modules/compute/schedmd-slurm-gcp-v5-node-group/source_image_logic.tf
Comment (cdunbar13) [reply community/modules/compute/schedmd-slurm-gcp-v5-node-group/source_image_logic.tf]: Updated

File: docs/vm-images.md
Comment (cboneti) [reply docs/vm-images.md]: is this resolved?

Comment (cdunbar13) [reply docs/vm-images.md]: Yes, resolved.

File: docs/vm-images.md
Comment (cdunbar13) [reply docs/vm-images.md]: Resolved.

**Review by cboneti — APPROVED** [review]

**Review by tpdownes — APPROVED** [review]

**Pull request merged**
